File: csharp/clock/Clock.cs
public class Clock
{
    private const int MaxValue = 24 * 60;
    private readonly int _minutes;
    
    public Clock(int hours, int minutes)
    {
        var corrected = hours * 60 + minutes;
        while (corrected < 0)
        {
            corrected += MaxValue;
        }
        _minutes = corrected % MaxValue;
    }

    public Clock Add(int minutesToAdd) => new Clock(0, _minutes + minutesToAdd);

    public Clock Subtract(int minutesToSubtract) => new Clock(0, _minutes - minutesToSubtract);
    
    public override string ToString()
    {
        var hours = (_minutes / 60) % 24;
        var minutes = _minutes % 60;
        return $"{hours:D2}:{minutes:D2}";
    }
    
    public override bool Equals(object obj)
    {
        if (obj is Clock other)
        {
            return _minutes == other._minutes;
        }

        return false;
    }

    public override int GetHashCode() => _minutes;
}

File: kotlin/clock/src/main/kotlin/Clock.kt
private const val MAX_MINUTES = 24 * 60

class Clock(hours: Int, minutes: Int) {

    private var minutes = (60 * hours + minutes).mod(MAX_MINUTES)

    fun subtract(minutes: Int) {
        this.minutes = (this.minutes - minutes).mod(MAX_MINUTES)
    }

    fun add(minutes: Int) {
        this.minutes = (this.minutes + minutes).mod(MAX_MINUTES)
    }

    override fun equals(other: Any?) = when (other) {
        is Clock -> other.minutes == this.minutes
        else -> false
    }

    override fun hashCode() = minutes

    override fun toString(): String {
        val displayHours = minutes / 60
        val displayMinutes = minutes % 60
        return "%02d:%02d".format(displayHours, displayMinutes)
    }
}
